at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart

```diff
@@ -0,0 +1,194 @@
+import 'dart:async';
+import 'dart:io';
+import 'package:at_commons/at_commons.dart';
+import 'package:at_lookup/src/cache/secondary_address_finder.dart';
+import 'package:at_lookup/src/util/lookup_util.dart';
+import 'package:at_utils/at_logger.dart';
+
+class CacheableSecondaryAddressFinder implements SecondaryAddressFinder {
+  static const Duration defaultCacheDuration = Duration(hours: 1);
+
+  final Map<String, SecondaryAddressCacheEntry> _map = {};
+  final _logger = AtSignLogger('SecondaryAddressCacheImpl');
+
+  final String _rootDomain;
+  final int _rootPort;
+  late SecondaryUrlFinder _secondaryFinder;
+
+  CacheableSecondaryAddressFinder(this._rootDomain, this._rootPort,
+      {SecondaryUrlFinder? secondaryFinder}) {
+    _secondaryFinder =
+        secondaryFinder ?? SecondaryUrlFinder(_rootDomain, _rootPort);
+  }
+
+  bool cacheContains(String atSign) {
+    return _map.containsKey(stripAtSignFromAtSign(atSign));
+  }
+
+  /// Returns the expiry time of this entry in millisecondsSinceEpoch, or null if this atSign is not in cache
+  int? getCacheExpiryTime(String atSign) {
+    atSign = stripAtSignFromAtSign(atSign);
+    if (_map.containsKey(atSign)) {
+      return _map[atSign]!.expiresAt;
+    } else {
+      return null;
+    }
+  }
+
+  @override
+  Future<SecondaryAddress> findSecondary(String atSign) async {
+    atSign = stripAtSignFromAtSign(atSign);
+
+    if (_cacheIsEmptyOrExpired(atSign)) {
+      // _updateCache will either populate the cache, or throw an exception
+      await _updateCache(atSign, defaultCacheDuration);
+    }
+    if (_map.containsKey(atSign)) {
+      // should always be true, since _updateCache will throw an exception if it fails
+      return _map[atSign]!.secondaryAddress;
+    } else {
+      // but just in case, we'll throw an exception if it's not
+      throw Exception(
+          "Failed to find secondary, in a theoretically impossible way");
+    }
+  }
+
+  bool _cacheIsEmptyOrExpired(String atSign) {
+    if (_map.containsKey(atSign)) {
+      SecondaryAddressCacheEntry entry = _map[atSign]!;
+      if (entry.expiresAt < DateTime.now().millisecondsSinceEpoch) {
+        // expiresAt is in the past - cache has expired
+        return true;
+      } else {
+        // cache has not yet expired
+        return false;
+      }
+    } else {
+      // cache is empty
+      return true;
+    }
+  }
+
+  static String stripAtSignFromAtSign(String atSign) {
```

Comment:
I thought we had a utility method already somewhere to do this. I believe this method can be moved to a reusable place.